src/ring_theory/localization.lean

```diff
@@ -79,302 +95,279 @@ localization, ring localization, commutative ring localization, characteristic p
-λ _ _ h, ext $ ring_hom.ext_iff.1 h
+variables (R)
 
-/-- Given `a : S`, `S` a localization of `R`, `is_integer a` iff `a` is in the image of
+-- TODO: define a subalgebra of `is_integer`s
+/-- Given `a : S`, `S` a localization of `R`, `is_integer R a` iff `a` is in the image of
 the localization map from `R` to `S`. -/
-def is_integer (a : S) : Prop := a ∈ set.range f.to_map
+def is_integer (a : S) : Prop := a ∈ (algebra_map R S).range
 
--- TODO: define a subalgebra of `is_integer`s
-lemma is_integer_zero : f.is_integer 0 := ⟨0, f.to_map.map_zero⟩
+end
 
-lemma is_integer_one : f.is_integer 1 := ⟨1, f.to_map.map_one⟩
+lemma is_integer_zero : is_integer R (0 : S) := subring.zero_mem _
+lemma is_integer_one : is_integer R (1 : S) := subring.one_mem _
 
-variables {f}
+lemma is_integer_add {a b : S} (ha : is_integer R a) (hb : is_integer R b) :
+  is_integer R (a + b) :=
+subring.add_mem _ ha hb
 
-lemma is_integer_add {a b} (ha : f.is_integer a) (hb : f.is_integer b) :
-  f.is_integer (a + b) :=
-begin
-  rcases ha with ⟨a', ha⟩,
-  rcases hb with ⟨b', hb⟩,
-  use a' + b',
-  rw [f.to_map.map_add, ha, hb]
-end
-
-lemma is_integer_mul {a b} (ha : f.is_integer a) (hb : f.is_integer b) :
-  f.is_integer (a * b) :=
-begin
-  rcases ha with ⟨a', ha⟩,
-  rcases hb with ⟨b', hb⟩,
-  use a' * b',
-  rw [f.to_map.map_mul, ha, hb]
-end
+lemma is_integer_mul {a b : S} (ha : is_integer R a) (hb : is_integer R b) :
+  is_integer R (a * b) :=
+subring.mul_mem _ ha hb
 
-lemma is_integer_smul {a : R} {b} (hb : f.is_integer b) :
-  f.is_integer (f.to_map a * b) :=
+lemma is_integer_smul {a : R} {b : S} (hb : is_integer R b) :
+  is_integer R (a • b) :=
 begin
   rcases hb with ⟨b', hb⟩,
   use a * b',
-  rw [←hb, f.to_map.map_mul]
+  rw [←hb, (algebra_map R S).map_mul, algebra.smul_def]
 end
 
-variables (f)
+variables (M)
 
 /-- Each element `a : S` has an `M`-multiple which is an integer.
 
 This version multiplies `a` on the right, matching the argument order in `localization_map.surj`.
 -/
 lemma exists_integer_multiple' (a : S) :
-  ∃ (b : M), is_integer f (a * f.to_map b) :=
-let ⟨⟨num, denom⟩, h⟩ := f.surj a in ⟨denom, set.mem_range.mpr ⟨num, h.symm⟩⟩
+  ∃ (b : M), is_integer R (a * algebra_map R S b) :=
+let ⟨⟨num, denom⟩, h⟩ := is_localization.surj _ a in ⟨denom, set.mem_range.mpr ⟨num, h.symm⟩⟩
 
 /-- Each element `a : S` has an `M`-multiple which is an integer.
 
 This version multiplies `a` on the left, matching the argument order in the `has_scalar` instance.
 -/
 lemma exists_integer_multiple (a : S) :
-  ∃ (b : M), is_integer f (f.to_map b * a) :=
-by { simp_rw mul_comm _ a, apply exists_integer_multiple' }
+  ∃ (b : M), is_integer R ((b : R) • a) :=
+by { simp_rw [algebra.smul_def, mul_comm _ a], apply exists_integer_multiple' }
 
-/-- Given `z : S`, `f.to_localization_map.sec z` is defined to be a pair `(x, y) : R × M` such
+/-- Given `z : S`, `is_localization.sec z` is defined to be a pair `(x, y) : R × M` such
```

Comment:
Wait, where is this `is_localization.sec`? An why does the statement below not use it? (It mentions `(to_localization_map M S).sec`.)

Reply:
This docstring was wrong, the correct spelling is `(to_localization_map M S).sec` as in the next docstring.

Reply:
As we discussed in the chat, `sec` deserves to be redefined for `is_localization`.